numpyro/contrib/module.py

```diff
@@ -433,3 +435,143 @@ def random_haiku_module(
+
+
+def random_nnx_module(
+    name,
+    nn_module,
+    prior,
+    mutable=None,
+):
+    """
+    A primitive to create a random :mod:`~flax.nnx` style neural network
+    which can be used in MCMC samplers. The parameters of the neural network
+    will be sampled from ``prior``.
+
+    :param str name: name of the module to be registered.
+    :param flax.nnx.Module nn_module: a pre-initialized `flax nnx` Module instance.
+    :param prior: a distribution or a dict of distributions or a callable.
+        If it is a distribution, all parameters will be sampled from the same
+        distribution. If it is a dict, it maps parameter names to distributions.
+        If it is a callable, it takes parameter name and parameter shape as
+        inputs and returns a distribution. For example::
+
+            class Linear(nnx.Module):
+                def __init__(self, din, dout, *, rngs):
+                    self.w = nnx.Param(jax.random.uniform(rngs.params(), (din, dout)))
+                    self.b = nnx.Param(jnp.zeros((dout,)))
+
+                def __call__(self, x):
+                    return x @ self.w + self.b
+
+            # Eager initialization
+            linear = Linear(din=4, dout=1, rngs=nnx.Rngs(params=random.PRNGKey(0)))
+            net = random_nnx_module("net", linear, prior={"w": dist.Normal(), "b": dist.Cauchy()})
+
+        Alternatively, we can use a callable. For example the following are equivalent::
+
+            prior=(lambda name, shape: dist.Cauchy() if name.endswith("b") else dist.Normal())
+            prior={"w": dist.Normal(), "b": dist.Cauchy()}
+
+    :param list mutable: A list to indicate mutable states of ``nn_module``. For example,
+        if your module has BatchNorm layer, we will need to define ``mutable=["batch_stats"]``.
+    :return: a callable that takes an array as an input and returns
+        the neural network transformed output array.
+    """
+    try:
+        from flax import nnx
+    except ImportError as e:
+        raise ImportError(
+            "Looking like you want to use flax.nnx to declare "
+            "nn modules. This is an experimental feature. "
+            "You need to install the latest version of `flax` to use this feature. "
+            "It can be installed with `pip install git+https://github.com/google/flax.git`."
+        ) from e
+
+    # Extract params and state directly from the module in one go
+    graph_def, params_state, _ = nnx.split(
+        nn_module, nnx.Param, nnx.filterlib.to_predicate(nnx.Not(nnx.Param))
+    )
+
+    param_dict = nnx.to_pure_dict(params_state)
+
+    numpyro.deterministic(f"{name}$params", param_dict)
+
+    def apply_fn(x, *fn_args, **fn_kwargs):
+        # Create a new model with the sampled parameters
+        graph_def, state = nnx.split(nn_module)
+        nnx.replace_by_pure_dict(state, param_dict)
+        model = nnx.merge(graph_def, state)
+        return model(x, *fn_args, **fn_kwargs)
+
+    return apply_fn
```

Comment:
I don't see how you set priors here. It might be better to use an implementation similar to flax/haiku (so that we can have a single logic to update from priors). No need to inspect the model again I guess. Just get the params structure from the output of `nnx_module` call, call _update_params to update it with the samples from priors. Then return the partial function.

Reply:
of course 🤦 ! Here is my first attempt https://github.com/pyro-ppl/numpyro/pull/1990/commits/4fe9f1133e06acd40017ec0403deba3b4a575aa5 wdyt?